File: backend/cloudwatch_metrics.py
"""CloudWatch Metrics client and analysis logic"""

import boto3
from datetime import datetime
from typing import List, Dict, Optional
from config import METRIC_CONFIGS, ANOMALY_THRESHOLD_SIGMA
import statistics


class CloudWatchMetricsService:
    """Service for querying and analyzing CloudWatch metrics"""
    
    def __init__(self, region: str = 'ap-south-1'):
        self.client = boto3.client('cloudwatch', region_name=region)
        self.region = region
    
    def get_metrics(
        self,
        service_type: str,
        resource_id: str,
        start_time: datetime,
        end_time: datetime
    ) -> List[Dict]:
        """Fetch CloudWatch metrics for a resource"""
        config = METRIC_CONFIGS.get(service_type)
        if not config:
            raise ValueError(f"Unsupported service type: {service_type}")
        
        period = self._calculate_period(start_time, end_time)
        metric_data_queries = []
        
        for idx, metric_name in enumerate(config['metrics']):
            metric_data_queries.append({
                'Id': f'm{idx}',
                'MetricStat': {
                    'Metric': {
                        'Namespace': config['namespace'],
                        'MetricName': metric_name,
                        'Dimensions': [
                            {
                                'Name': config['dimension_name'],
                                'Value': resource_id
                            }
                        ]
                    },
                    'Period': period,
                    'Stat': 'Average'
                }
            })
        
        try:
            response = self.client.get_metric_data(
                MetricDataQueries=metric_data_queries,
                StartTime=start_time,
                EndTime=end_time
            )
            
            results = []
            for idx, result in enumerate(response.get('MetricDataResults', [])):
                metric_name = config['metrics'][idx]
                
                values = [
                    {
                        'timestamp': ts.isoformat(),
                        'value': val
                    }
                    for ts, val in zip(result.get('Timestamps', []), result.get('Values', []))
                ]
                
                analysis = self._analyze_metric([v['value'] for v in values])
                
                metric_analysis = {
                    'resourceId': resource_id,
                    'metricName': metric_name,
                    'namespace': config['namespace'],
                    'values': values,
                    'anomalyDetected': analysis['anomaly_detected']
                }
                
                if analysis.get('threshold_breached'):
                    metric_analysis['thresholdBreached'] = analysis['threshold_breached']
                
                results.append(metric_analysis)
            
            return results
            
        except self.client.exceptions.ClientError as e:
            error_code = e.response['Error']['Code']
            if error_code == 'AccessDeniedException':
                raise PermissionError('Insufficient permissions to access CloudWatch metrics')
            raise
    
    def _calculate_period(self, start_time: datetime, end_time: datetime) -> int:
        """Calculate appropriate period based on time range"""
        hours = (end_time - start_time).total_seconds() / 3600
        return 300 if hours <= 24 else 3600  # 5 minutes for <24h, 1 hour for >24h
    
    def _analyze_metric(self, values: List[float]) -> Dict:
        """Analyze metric values for anomalies"""
        if not values:
            return {'anomaly_detected': False}
        
        mean = statistics.mean(values)
        
        if len(values) < 2:
            return {'anomaly_detected': False}
        
        stdev = statistics.stdev(values)
        threshold = mean + (ANOMALY_THRESHOLD_SIGMA * stdev)
        
        for idx, value in enumerate(values):
            if value > threshold:
                return {
                    'anomaly_detected': True,
                    'threshold_breached': {
                        'threshold': threshold,
                        'breachedAt': datetime.utcnow().isoformat()
                    }
                }
        
        return {'anomaly_detected': False}

File: backend/src/cloudwatchMetrics.ts
// CloudWatch Metrics client and analysis logic

import {
  CloudWatchClient,
  GetMetricDataCommand,
  MetricDataQuery,
  MetricStat
} from '@aws-sdk/client-cloudwatch';
import { MetricAnalysis, ServiceType } from './types.js';
import { METRIC_CONFIGS, ANOMALY_THRESHOLD_SIGMA } from './config.js';

export class CloudWatchMetricsService {
  private client: CloudWatchClient;
  
  constructor(region: string = 'ap-south-1') {
    this.client = new CloudWatchClient({ region });
  }
  
  async getMetrics(
    serviceType: ServiceType,
    resourceId: string,
    startTime: Date,
    endTime: Date
  ): Promise<MetricAnalysis[]> {
    const config = METRIC_CONFIGS[serviceType];
    if (!config) {
      throw new Error(`Unsupported service type: ${serviceType}`);
    }
    
    const queries: MetricDataQuery[] = config.metrics.map((metricName, index) => ({
      Id: `m${index}`,
      MetricStat: {
        Metric: {
          Namespace: config.namespace,
          MetricName: metricName,
          Dimensions: [
            {
              Name: config.dimensionName,
              Value: resourceId
            }
          ]
        },
        Period: this.calculatePeriod(startTime, endTime),
        Stat: 'Average'
      } as MetricStat
    }));
    
    try {
      const command = new GetMetricDataCommand({
        MetricDataQueries: queries,
        StartTime: startTime,
        EndTime: endTime
      });
      
      const response = await this.client.send(command);
      
      if (!response.MetricDataResults) {
        return [];
      }
      
      return response.MetricDataResults.map((result, index) => {
        const metricName = config.metrics[index];
        const values = (result.Timestamps || []).map((timestamp, i) => ({
          timestamp: timestamp.toISOString(),
          value: result.Values?.[i] || 0
        }));
        
        const analysis = this.analyzeMetric(values.map(v => v.value));
        
        return {
          resourceId,
          metricName,
          namespace: config.namespace,
          values,
          anomalyDetected: analysis.anomalyDetected,
          thresholdBreached: analysis.thresholdBreached
        };
      });
    } catch (error: any) {
      if (error.name === 'AccessDeniedException') {
        throw new Error('Insufficient permissions to access CloudWatch metrics');
      }
      throw error;
    }
  }
  
  private calculatePeriod(startTime: Date, endTime: Date): number {
    const hours = (endTime.getTime() - startTime.getTime()) / (1000 * 60 * 60);
    return hours <= 24 ? 300 : 3600; // 5 minutes for <24h, 1 hour for >24h
  }
  
  private analyzeMetric(values: number[]): {
    anomalyDetected: boolean;
    thresholdBreached?: { threshold: number; breachedAt: string };
  } {
    if (values.length === 0) {
      return { anomalyDetected: false };
    }
    
    const mean = values.reduce((sum, val) => sum + val, 0) / values.length;
    const variance = values.reduce((sum, val) => sum + Math.pow(val - mean, 2), 0) / values.length;
    const stdDev = Math.sqrt(variance);
    
    const threshold = mean + (ANOMALY_THRESHOLD_SIGMA * stdDev);
    
    for (let i = 0; i < values.length; i++) {
      if (values[i] > threshold) {
        return {
          anomalyDetected: true,
          thresholdBreached: {
            threshold,
            breachedAt: new Date(Date.now() - (values.length - i) * 5 * 60 * 1000).toISOString()
          }
        };
      }
    }
    
    return { anomalyDetected: false };
  }
}
